xlive/H2MOD/Modules/WeaponOffsets/WeaponOffsets.cpp

```diff
@@ -0,0 +1,218 @@
+			{{0.02, 0, 0}, "objects\\weapons\\rifle\\smg\\smg", {0,0,0}, NULL},
+			{{0.01, 0, 0}, "objects\\weapons\\rifle\\sniper_rifle\\sniper_rifle", {0,0,0}, NULL}
+		};
+
+		namespace
+		{
+			std::map<int, std::map<e_weapon_offsets_string, char*>> string_table;
+			// Used for controls that use the same string, A identifier has to be appended to them
+			// I.E Reset##1... Reset##20
+			std::map<std::string, std::string> string_cache;
+			void OffsetMenu(Weapons weapon, char* slider, e_weapon_offsets_string text, float& offset, const float default_value)
+			{
+				// Bullshit for unique widget ids
+				char item2[32], item3[32];
+				strcpy(item2, slider);
+				strcat(item2, "1");
+				strcpy(item3, slider);
+				strcat(item3, "2");
+
+				ImVec2 item_size = ImGui::GetItemRectSize();
+				ImVec2 b2_size = ImVec2(WidthPercentage(20), item_size.y);
+
+				ImGui::Text(GetString(text));
+				ImGui::PushItemWidth(WidthPercentage(60));
+				ImGui::SliderFloat(slider, &offset, -0.15, 0.15, ""); ImGui::SameLine();
+				if (ImGui::IsItemEdited() && customOffsets[weapon].tag != nullptr) { ApplyOffset(weapon); }
+
+				ImGui::PushItemWidth(WidthPercentage(20));
+				ImGui::InputFloat(item2, &offset, -0.15, 0.15, "%.3f"); ImGui::SameLine();
+				if (ImGui::IsItemEdited() && customOffsets[weapon].tag != nullptr) { ApplyOffset(weapon); }
+
+				ImGui::PushItemWidth(WidthPercentage(20));
+				if (ImGui::Button(GetString(reset, item3), b2_size))
+				{
+					offset = default_value;
+					if (customOffsets[weapon].tag != nullptr) { ApplyOffset(weapon); }
+				}
+				ImGui::PopItemWidth();
+			}
+			void OffsetSettings()
+			{
+
+				static const char* weapons[]
+				{ GetString(battle_rifle_title), GetString(beam_rifle_title),GetString(brute_plasma_rifle_title), GetString(brute_shot_title),
+				GetString(carbine_title), GetString(energy_sword_title), GetString(fuel_rod_title), GetString(magnum_title),
+				GetString(needler_title), GetString(plasma_pistol_title), GetString(plasma_rifle_title), GetString(rocket_launcher_title),
+				GetString(sentinel_beam_title), GetString(shotgun_title), GetString(smg_title), GetString(sniper_title) };
+				static int selectedOption = 0;
+
+				// Setup combo box menus for each weapon
+				ImGui::Combo(GetString(combo_title), &selectedOption, weapons, IM_ARRAYSIZE(weapons));
+				OffsetMenu(static_cast<Weapons>(selectedOption), "##OffsetX", weapon_offset_x,
+					customOffsets[selectedOption].ModifiedOffset.i, customOffsets[selectedOption].DefaultOffset.i);
+				OffsetMenu(static_cast<Weapons>(selectedOption), "##OffsetY", weapon_offset_y,
+					customOffsets[selectedOption].ModifiedOffset.j, customOffsets[selectedOption].DefaultOffset.j);
+				OffsetMenu(static_cast<Weapons>(selectedOption), "##OffsetZ", weapon_offset_z,
+					customOffsets[selectedOption].ModifiedOffset.k, customOffsets[selectedOption].DefaultOffset.k);
+			}
+		}
+
+		char* GetString(e_weapon_offsets_string string, const std::string& id)
```

Comment:
if this function returns a string buffer that shouldn't be modified, it should return `const char*`. I noticed you just copied the code from another place expecting it to be fine. you might want to fix this there as well and remove the casting that removes the constness of the string buffers.